### PR TITLE
SX1272 LoRa radio support

### PR DESCRIPTION
**Pull request by tagunil**
This is an alternative to #29085. I've attempted to prevent code duplication by reworking existing SX1276 driver towards SX1272 support. Feedback is very welcome.
I've tested this driver on several custom STM32L1+SX1272/SX1276-based devices, but further testing would be also much appreciated.

### CODE REVIEW

**Comment by tagunil** [comment]
@mniestroj @andysan @Mani-Sadhasivam Could you look at this and verify my approach?

**Comment by greg-leach** [comment]
I can provide two products suitable for testing this - please let me know if you'd like it sent to you.

**Comment by tagunil** [comment]
> I can provide two products suitable for testing this - please let me know if you'd like it sent to you.

Thank you very much, but I've already got some HW. Now I'm trying to find some time for testing.

File: drivers/lora/sx127x.c [hunk 1]
```diff
@@ -331,94 +400,100 @@ static int sx1276_transceive(uint8_t reg, bool write, void *data, size_t length)
...
+#ifdef RF_PACONFIG_MAX_POWER_MASK
 		power = clamp_int8(power, -4, 15);
 
 		pa_dac |= RF_PADAC_20DBM_OFF;
 		if (power > 0) {
 			/* Set the power range to 0 -- 10.8+0.6*7 dBm. */
-			pa_config |= 7 << SX1276_PA_CONFIG_MAX_POWER_SHIFT;
-			pa_config |= (power & 0x0F);
+			pa_config |= ~RF_PACONFIG_MAX_POWER_MASK;
+			pa_config |= (power & RF_PACONFIG_OUTPUTPOWER_MASK);
```
Comment (mniestroj) [inline drivers/lora/sx127x.c:472]: `RF_PACONFIG_OUTPUTPOWER_MASK` is `0xF0`, but we used `0x0F` so far. Also `pa_config |= ~RF_PACONFIG_MAX_POWER_MASK;` line seems suspicious. Those mask macros are already inverted in loramac-node, which looks odd to me. The usual pattern is to do `reg &= ~mask;`, `reg |= bits;` and `reg |= (val & mask);`. What about creating helper macros like:
```
#define SX127X_PACONFIG_MAX_POWER_MASK (~RF_PACONFIG_MAX_POWER_MASK)
#define SX127X_PACONFIG_OUTPUTPOWER_MASK (~RF_PACONFIG_OUTPUTPOWER_MASK)
```
and use them with the patterns above?

Comment (mniestroj) [reply drivers/lora/sx127x.c:472]: A better alternative from Zephyr point of view would be to post changes (inversion) of `RF_PACONFIG_MAX_POWER_MASK` and `RF_PACONFIG_OUTPUTPOWER_MASK` to mainline loramac-node. But that could potentially harm other projects. Maybe it is good to try and see what is the feedback there.

Comment (tagunil) [reply drivers/lora/sx127x.c:472]: Nice catch, thanks! These masks are weird and confusing, but I'm afraid they're public API for a long time. I'll create the helpers for now.

Comment (tagunil) [reply drivers/lora/sx127x.c]: I've wrapped these RF_ macros to the helper ones with more usual semantics. @mniestroj, could you please look again?

**Comment by chrta** [comment]
I tested this on hardware with #29841 and a nucleo_f429zi board. Since i do not have a lorawan gateway, i only saw that the driver is initialized and i can send a lora packet at the specified frequency and i can send a cw signal. Both verified with an sdr by inspecting the radio spectrum.
-> It seems that this code is working.

**Comment by tagunil** [comment]
> I tested this on hardware with #29841 and a nucleo_f429zi board. Since i do not have a lorawan gateway, i only saw that the driver is initialized and i can send a lora packet at the specified frequency and i can send a cw signal. Both verified with an sdr by inspecting the radio spectrum.
> -> It seems that this code is working.

Thanks! I do have a gateway, so I've tested Join procedure and data uplink over a SX1272.

**Review by Mani-Sadhasivam — APPROVED** [review]
LGTM! Thanks for your patience and continuous effort.

**Review by mniestroj — APPROVED** [review]

**Comment by Mani-Sadhasivam** [comment]
@carlescufi @nashif Please merge!

**Pull request merged**